lib/handoff/transfer_list.c

```diff
@@ -0,0 +1,334 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Linaro Limited
+ */
+
+#include <assert.h>
+#include <common/debug.h>
+#include <inttypes.h>
+#include <lib/transfer_list.h>
+#include <lib/utils_def.h>
+#include <string.h>
+
+#define ADD_OVERFLOW(a, b, res) __builtin_add_overflow((a), (b), (res))
+
+#define SUB_OVERFLOW(a, b, res) __builtin_sub_overflow((a), (b), (res))
+
+#define MUL_OVERFLOW(a, b, res) __builtin_mul_overflow((a), (b), (res))
+
+#define ROUNDUP_OVERFLOW(v, size, res) (__extension__({ \
+	typeof(*(res)) __roundup_tmp = 0; \
+	typeof(v) __roundup_mask = (typeof(v))(size) - 1; \
+	\
+	ADD_OVERFLOW((v), __roundup_mask, &__roundup_tmp) ? 1 : \
+		(void)(*(res) = __roundup_tmp & ~__roundup_mask), 0; \
+}))
+
+// 2.4.5 Creating a TL
+struct transfer_list_header *transfer_list_init(void *p, size_t max_size)
+{
+	uintptr_t mask = (1 << TRANSFER_LIST_INIT_MAX_ALIGN) - 1;
+	struct transfer_list_header *tl = p;
+
+	if (((uintptr_t)p & mask) || (max_size & mask) ||
+	    max_size < sizeof(*tl))
+		return NULL;
+
+	memset(tl, 0, max_size);
+	tl->signature = TRANSFER_LIST_SIGNATURE;
+	tl->version = TRANSFER_LIST_VERSION;
+	tl->hdr_size = sizeof(*tl);
+	tl->alignment = TRANSFER_LIST_INIT_MAX_ALIGN; // initial max align
+	tl->size = sizeof(*tl); // initial size is the size of header
+	tl->max_size = max_size;
+	transfer_list_update_checksum(tl);
+	return tl;
+}
+
+// 2.4.6 Relocating a TL
+bool transfer_list_relocate(struct transfer_list_header *tl, void **p, size_t max_size)
+{
+	if (!tl || !p || !(*p) || max_size == 0)
+		return false;
+
+	uintptr_t mask = (1 << tl->alignment) - 1;
+	uintptr_t new_tl_base = ((uintptr_t)(*p) + mask) & ~mask;
+
+	if (new_tl_base < (uintptr_t)(*p))
+		new_tl_base += mask + 1;
+
+	if (new_tl_base - (uintptr_t)(*p) + tl->size > max_size)
+		return false;
+
+	memmove((void *)new_tl_base, (void *)tl, tl->size);
+	((struct transfer_list_header *)new_tl_base)->max_size = max_size; //to-check
+	*p = (void *)new_tl_base;
+	transfer_list_update_checksum((struct transfer_list_header *)(*p));
+	return true;
+}
+
+static bool check_header(struct transfer_list_header *tl)
+{
+	if (!tl)
+		return false;
+
+	if (tl->signature != TRANSFER_LIST_SIGNATURE) {
+		ERROR("Bad transfer list signature %#"PRIx32"\n",
+		      tl->signature);
+		return false;
+	}
+	if (tl->version != TRANSFER_LIST_VERSION) {
+		ERROR("Unsupported transfer list version %#"PRIx32"\n",
+		      tl->version);
+		return false;
+	}
+	if (!tl->max_size) {
+		ERROR("Bad transfer list max size %#"PRIx32"\n",
+		      tl->max_size);
+		return false;;
+	}
+	if (tl->size > tl->max_size) {
+		ERROR("Bad transfer list size %#"PRIx32"\n", tl->size);
+		return false;;
+	}
+	return true;
+}
+
+// 2.4.1 Validating a TL header
+struct transfer_list_header *transfer_list_check_header(struct transfer_list_header *tl)
```

Comment:
Since you aren't changing the TL here, this can be a bool function

Reply:
Fixed.